sound/soc/sof/intel/hda.c

```diff
@@ -1045,12 +1045,32 @@ static bool link_slaves_found(struct snd_sof_dev *sdev,
 			 */
 			unique_id = SDW_UNIQUE_ID(adr);
 			if (link->num_adr == 1 ||
-			    ids[j].id.unique_id == SDW_IGNORED_UNIQUE_ID ||
 			    ids[j].id.unique_id == unique_id) {
 				dev_dbg(bus->dev,
 					"found %x at link %d\n",
 					part_id, link_id);
 				break;
+			} else if (ids[j].id.unique_id == SDW_IGNORED_UNIQUE_ID) {
+				unsigned int part_id2;
+
+				/* make sure there is no the same codec on the link */
+				for (k = 0; k < link->num_adr; k++) {
+					if (k == i)
+						continue;
+					part_id2 = SDW_PART_ID(link->adr_d[k].adr);
+					if (part_id2 == part_id) {
+						dev_dbg(bus->dev,
+							"Slave %x id %d not found\n",
+							part_id, unique_id);
+						return false;
+					}
+				}
+				if (k == link->num_adr) {
+					dev_dbg(bus->dev,
+						"found %x at link %d\n",
+						part_id, link_id);
+					break;
+				}
```

Comment:
I don't think this fully works in all cases.

Take an hypothetical case with four amps on a link, and a machine only has 2 of those amps (matching the unique ID of the 4 amp case)
If the first descriptor has 4 links, it will be selected.

I think we are missing a test that the total number of devices in the machine descriptor matches the total number of slaves detected.

Reply:
> I don't think this fully works in all cases.
> 
> Take an hypothetical case with four amps on a link, and a machine only has 2 of those amps (matching the unique ID of the 4 amp case)
> If the first descriptor has 4 amps on a link, it will be selected.
> 
> I think we are missing a test that the total number of devices in the machine descriptor matches the total number of slaves detected.

@plbossart Yes, we can't cover all cases if we don't test the total number. But on the other hand, we will lose the tolerance when BIOS has some additional _ADRs. For example, rt700 and rt701 can be seen on almost all products' BIOS. We currently can ignore those superfluous _ADRs. 
I do not count rt700 and rt701 on this PR version, but we have to think a better way to deal with those superfluous _ADRs.